Source/DafnyDriver/LocalTestLoggerEvents.cs

```diff
@@ -0,0 +1,423 @@
+      string traceDisplayName) {
+      ValidateArg.NotNull(eventHandlersFactory, nameof(eventHandlersFactory));
+      ValidateArg.NotNull(args, nameof(args));
+
+      // Invoke the handlers on a background thread.
+      this.loggerEventQueue.QueueJob(
+        () => {
+          var eventHandlers = eventHandlersFactory();
+          eventHandlers?.SafeInvoke(this, args, traceDisplayName);
+        }, size);
+    }
+
+    /// <summary>
+    /// Method called to process a job which is coming from the logger event queue.
+    /// </summary>
+    private void ProcessQueuedJob(Action action) {
+      action();
+    }
+
+    /// <summary>
+    /// Throws if we are disposed.
+    /// </summary>
+    private void CheckDisposed() {
+      if (this.isDisposed) {
+        throw new ObjectDisposedException(typeof(TestLoggerEvents).FullName);
+      }
+    }
+
+    /// <summary>
+    /// The method parses the config file of vstest.console.exe to see if the Max Job Queue Length is defined.
+    /// Return the Max Queue Length so defined or a default value specified by TestPlatformDefaults.DefaultMaxLoggerEventsToCache
+    /// </summary>
+    private int GetMaxNumberOfJobsInQueue() {
+      return GetSetting(TestPlatformDefaults.MaxNumberOfEventsLoggerEventQueueCanHold,
+        TestPlatformDefaults.DefaultMaxNumberOfEventsLoggerEventQueueCanHold);
+    }
+
+    /// <summary>
+    /// The method parses the config file of vstest.console.exe to see if the Max Job Queue size is defined.
+    /// Return the Max Queue size so defined or a default value specified by TestPlatformDefaults.DefaultMaxJobQueueSize
+    /// </summary>
+    private int GetMaxBytesQueueCanHold() {
+      return GetSetting(TestPlatformDefaults.MaxBytesLoggerEventQueueCanHold,
+        TestPlatformDefaults.DefaultMaxBytesLoggerEventQueueCanHold);
+    }
+
+    /// <summary>
+    /// Returns whether flow control on logger events queue should be enabled or not. Default is enabled.
+    /// </summary>
+    private static bool IsBoundsEnabledOnLoggerEventQueue() {
+      bool enableBounds;
+#if NETFRAMEWORK
+            string enableBoundsOnEventQueueIsDefined =
+ ConfigurationManager.AppSettings[TestPlatformDefaults.EnableBoundsOnLoggerEventQueue];
+#else
+      string enableBoundsOnEventQueueIsDefined = null;
+#endif
+      if (string.IsNullOrEmpty(enableBoundsOnEventQueueIsDefined)) {
+        enableBounds = TestPlatformDefaults.DefaultEnableBoundsOnLoggerEventQueue;
+      } else {
+        if (!(bool.TryParse(enableBoundsOnEventQueueIsDefined, out enableBounds))) {
+          enableBounds = TestPlatformDefaults.DefaultEnableBoundsOnLoggerEventQueue;
+        }
+      }
+
+      return enableBounds;
+    }
+
+    /// <summary>
+    /// Returns the approximate size of a TestResult instance.
+    /// </summary>
+    private static int FindTestResultSize(TestResultEventArgs args) {
+      Debug.Assert(args != null && args.Result != null);
```

Comment:
Elsewhere in Dafny, we use `Contract.Assert`. Is this an intentional step to move us from `Contract.` to `Debug.`?

Also, if you're okay with `Contract.`, then, since this is a precondition, it can be written `Contract.Requires(...);`, which expresses that it's the caller's responsibility to live up to the condition. (The disadvantage of `Contract.Requires` is that we no longer are getting run-time checks for it.)

Reply:
Ditto.